src/matrix/kaldi-matrix.cc

```diff
@@ -2784,6 +2784,78 @@ void MatrixBase<Real>::DiffTanh(const MatrixBase<Real> &value,
 
 
 template<typename Real>
+void MatrixBase<Real>::AddSpatialRegularizationDeriv(
+    const MatrixBase<Real>& out_value, Real scale, Real* regularization_sumsq) {
+  const unsigned int kBlockSize = 256;
+  const unsigned int kSectWidth = 16;
+  KALDI_ASSERT(SameDim(*this, out_value));
+
+  int ext_cols = ((this->NumCols() + (kSectWidth - 1)) / kSectWidth)
+      * kSectWidth;
+  Matrix<Real> val_ext(this->NumRows(), ext_cols);
+  Matrix<Real> conv1(this->NumRows(), ext_cols);
+  Matrix<Real> conv2(this->NumRows(), ext_cols);
+
+  // add 0 padding to the end of the rows
+  for (int32 i = 0; i < this->NumRows(); i++) {
+    for (int32 j = 0; j < this->NumCols(); j++) {
+      val_ext(i, j) = out_value(i, j);
```

Comment:
you can do this copying with a submatrix, e.g.
val_ext.RowRange(0, out_value.NumCols()).CopyFromMat(out_value);